readme.md

```diff
@@ -1,848 +1,1214 @@
+		return;
+	}
 
+	var foo = doSomethingSlow();
 
-    // 2.B.2.1
-    // Named Function Declaration
-    function foo( arg1, argN ) {
+	// some statements
+}
+```
 
-    }
+##### 2.2.2.3. Function declaration
 
-    // Usage.
-    foo( arg1, argN );
+Prefer function declaration instead of function expression to avoid issues with executing function before it has been defined.
 
+```javascript
+// Good
+function square( number ) {
+	return number * number;
+}
 
-    // 2.B.2.2
-    // Named Function Declaration.
-    function square( number ) {
-        return number * number;
-    }
+square( 10 );
 
-    // Usage.
-    square( 10 );
+// Bad
+var square = function( number ) {
+	return number * number;
+};
 
-    // Really contrived continuation passing style.
-    function square( number, callback ) {
-        callback( number * number );
-    }
+square( 10 );
+```
 
-    square( 10, function( square ) {
-        // callback statements
-    } );
+##### 2.2.2.4. Continuation-Passing style
 
+If your function should be executed asynchronously or requires additional callback (e.g. filter function) as an function argument, pass it as the last argument in function signature. Prefer [Promises](#2227-promises) for asynchronous code when possible.
 
-    // 2.B.2.3
-    // Function Expression
-    var square = function( number ) {
-        // Return something valuable and relevant.
-        return number * number;
-    };
+```javascript
+// Declaration.
+function square( number, callback ) {
+	callback( number * number );
+}
 
+// Usage.
+square( 10, function( square ) {
+	// callback statements
+} );
+```
 
-    // 2.B.2.4
-    // Constructor Declaration
-    function FooBar( options ) {
-        this.options = options;
-    }
+##### 2.2.2.5. Promises
 
-    // Usage.
-    var fooBar = new FooBar( { a: 'alpha' } );
+Preferably, asynchronous code is written using Promises. It's not always possible when maintaining code, however, prefer Promise based code for new features and everythere where you have control over API structure consumer. Always use `CKEDITOR.tools.promise` instead of native `Promise` object for wider browser support.
 
-    fooBar.options; // -> { a: 'alpha' }
+```javascript
+var promise = new CKEDITOR.tools.promise( function( resolve, reject ) {
+	var result = getContentFromFileServer( options );
 
-    ```
+	if ( result.status === 200 ) {
+		resolve( result.data );
+	} else {
+		reject( result.error );
+	}
+} );
+```
 
+##### 2.2.2.6 Constructor declaration
 
-    C. Quotes
+Prefer using [CKEDITOR.tools.createClass](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_tools.html#method-createClass) API instead of native constructor declaration.
 
-    Always use single quotes in JavaScript, but double quotes in HTML.
+```javascript
+// Good
+var FooBar = CKEDITOR.tools.createClass( {
+	$: function( message ) {
+		this.message = message;
+	},
+	
+	proto: {
+		greeting: function() {
+			console.log( this.message );
+		}
+	}
```

Comment:
`createClass` supports also privates, inheritance and static methods. Maybe we could show an example for it as well. It doesn't have to be the same example, there could be multiple examples or code references showing different usages of this tool.
https://github.com/ckeditor/ckeditor4/blob/411985373e833a0df76e4ebd0afb45f337612725/core/tools.js#L759-L763

Reply:
I'm not sure about that. The main information seems to be included i.e. use `createClass` with API reference, do we need copy docs here?